File: clients/julia/src/compile.jl

function get_make()
    get(ENV, "MAKE", "make")
end

function validate_stan_dir(path::AbstractString)
    if !isdir(path)
        error("Path does not exist!\n$path")
    end
    if !isfile(joinpath(path, "Makefile"))
        error(
            "Makefile does not exist at path! Make sure it was installed correctly.\n$path",
        )
    end
end

"""
    set_tinystan_path!(path)

Set the path TinyStan.
"""
function set_tinystan_path!(path::AbstractString)
    validate_stan_dir(path)
    ENV["TINYSTAN"] = path
end

"""
    get_tinystan_path() -> String

Return the path the the TinyStan directory.

If the environment variable `TINYSTAN` is set, this will be returned.
Otherwise, this function downloads a matching version of TinyStan under
a folder called `.tinystan` in the user's home directory.

See [`set_tinystan_path!()`](@ref) to set the path from within Julia.
"""
function get_tinystan_path()
    path = get(ENV, "TINYSTAN", "")
    if path == ""
        path = CURRENT_TINYSTAN
        try
            validate_stan_dir(path)
        catch
            println(
                "TinyStan not found at location specified by \$TINYSTAN " *
                "environment variable, downloading version $pkg_version to $path",
            )
            get_tinystan_src()
            num_files = length(readdir(HOME_TINYSTAN))
            if num_files >= 5
                @warn "Found $num_files different versions of TinyStan in $HOME_TINYSTAN. " *
                      "Consider deleting old versions to save space."
            end
            println("Done!")
        end
    end
    return path
end

"""
    compile_model(stan_file; stanc_args=[], make_args=[])

Run TinyStan’s Makefile on a `.stan` file, creating the `.so` used by StanModel and
return a path to the compiled library.
Arguments to `stanc3` can be passed as a vector, for example `["--O1"]` enables level 1 compiler
optimizations.
Additional arguments to `make` can be passed as a vector, for example `["STAN_NO_RANGE_CHECKS=true"]`
will disable bounds checking in the Stan Math library. If the same flags are defined in `make/local`,
the versions passed here will take precedent.

This function checks that the path to TinyStan is valid and will error if it is not.
This can be set with [`set_tinystan_path!()`](@ref).
"""
function compile_model(
    stan_file::AbstractString;
    stanc_args::AbstractVector{String} = String[],
    make_args::AbstractVector{String} = String[],
)
    tinystan = get_tinystan_path()
    validate_stan_dir(tinystan)

    if !isfile(stan_file)
        throw(SystemError("Stan file not found: $stan_file"))
    end
    if splitext(stan_file)[2] != ".stan"
        error("File '$stan_file' does not end in .stan")
    end

    absolute_path = abspath(stan_file)
    output_file = splitext(absolute_path)[1] * "_model.so"

    cmd = Cmd(
        `$(get_make()) $make_args "STANCFLAGS=--include-paths=. $stanc_args" $output_file`,
        dir = abspath(tinystan),
    )
    out = IOBuffer()
    err = IOBuffer()
    is_ok = success(pipeline(cmd; stdout = out, stderr = err))
    if !is_ok
        error(
            "Compilation failed!\nCommand: $cmd\nstdout: $(String(take!(out)))\nstderr: $(String(take!(err)))",
        )
    end
    return output_file
end

WINDOWS_PATH_SET = Ref{Bool}(false)

function tbb_found()
    try
        run(pipeline(`where.exe tbb.dll`, stdout = devnull, stderr = devnull))
    catch
        return false
    end
    return true
end

function windows_dll_path_setup()
    if Sys.iswindows() && !(WINDOWS_PATH_SET[])
        if tbb_found()
            WINDOWS_PATH_SET[] = true
        else
            # add TBB to %PATH%
            ENV["PATH"] =
                joinpath(get_tinystan_path(), "stan", "lib", "stan_math", "lib", "tbb") *
                ";" *
                ENV["PATH"]
            WINDOWS_PATH_SET[] = tbb_found()
        end
    end
end

File: clients/R/R/compile.R
IS_WINDOWS <- isTRUE(.Platform$OS.type == "windows")
MAKE <- Sys.getenv("MAKE", "make")


verify_tinystan_path <- function(path) {
  suppressWarnings({
    folder <- normalizePath(path)
  })
  if (!dir.exists(folder)) {
    stop(paste0(
      "TinyStan folder '",
      folder,
      "' does not exist!\n",
      "If you need to set a different location, call 'set_tinystan_path()'"
    ))
  }
  makefile <- file.path(folder, "Makefile")
  if (!file.exists(makefile)) {
    stop(paste0(
      "TinyStan folder '",
      folder,
      "' does not contain file 'Makefile',",
      " please ensure it is built properly!\n",
      "If you need to set a different location, call 'set_tinystan_path()'"
    ))
  }
}

#' @title Function `set_tinystan_path()`
#' @description Set the path to TinyStan.
#' @details This should point to the top-level folder of the repository.
#' @export
set_tinystan_path <- function(path) {
  verify_tinystan_path(path)
  Sys.setenv(TINYSTAN = normalizePath(path))
}

#' Get the path to TinyStan.
#'
#' By default this is set to the value of the environment
#' variable `TINYSTAN`.
#'
#' If there is no path set, this function will download
#' a matching version of TinyStan to a folder called
#' `.tinystan` in the user's home directory.
#'
#' @seealso [set_tinystan_path]
get_tinystan_path <- function() {
  # try to get from environment
  path <- Sys.getenv("TINYSTAN", unset = "")
  if (path == "") {
    path <- CURRENT_TINYSTAN
    tryCatch(
      {
        verify_tinystan_path(path)
      },
      error = function(e) {
        print(paste0(
          "TinyStan not found at location specified by $TINYSTAN ",
          "environment variable, downloading version ",
          packageVersion("tinystan"),
          " to ",
          path
        ))
        get_tinystan_src()
      }
    )
  }

  return(path)
}


#' @title Function `compile_model()`
#' @description Compiles a Stan model.
#' @details Run TinyStan's Makefile on a `.stan` file, creating
#' the `.so` used by the StanModel class.
#' This function checks that the path to TinyStan is valid
#' and will error if not. This can be set with `set_tinystan_path`.
#'
#' @param stan_file A path to a Stan model file.
#' @param stanc_arg A vector of arguments to pass to stanc3.
#' For example, `c('--O1')` will enable compiler optimization level 1.
#' @param make_args A vector of additional arguments to pass to Make.
#' For example, `c('STAN_NO_RANGE_CHECKS=True')` will disable bounds
#' checking in the Stan Math library. If the same flags are defined
#' in `make/local`, the versions passed here will take precedent.
#' @return Path to the compiled model.
#'
#' @seealso [tinystan::set_tinystan_path()]
#' @export
compile_model <- function(stan_file, stanc_args = NULL, make_args = NULL) {
  verify_tinystan_path(get_tinystan_path())
  suppressWarnings({
    file_path <- normalizePath(stan_file)
  })
  if (tools::file_ext(file_path) != "stan") {
    stop(paste0("File '", file_path, "' does not end with '.stan'"))
  }
  if (!file.exists(file_path)) {
    stop(paste0("File '", file_path, "' does not exist!"))
  }

  output <- paste0(tools::file_path_sans_ext(file_path), "_model.so")
  stancflags <- paste("--include-paths=.", paste(stanc_args, collapse = " "))

  flags <- c(
    paste("-C", get_tinystan_path()),
    make_args,
    paste0("STANCFLAGS=\"", stancflags, "\""),
    output
  )

  suppressWarnings({
    res <- system2(MAKE, args = flags, stdout = TRUE, stderr = TRUE)
  })
  res_attrs <- attributes(res)
  if ("status" %in% names(res_attrs) && res_attrs$status != 0) {
    stop(paste0(
      "Compilation failed with error code ",
      res_attrs$status,
      "\noutput:\n",
      paste(res, collapse = "\n")
    ))
  }

  return(output)
}

tbb_found <- function() {
  suppressWarnings(
    out <- system2("where.exe", "tbb.dll", stdout = NULL, stderr = NULL)
  )
  return(out == 0)
}

WINDOWS_PATH_SET <- FALSE

windows_dll_path_setup <- function() {
  if (.Platform$OS.type == "windows" && !WINDOWS_PATH_SET) {
    if (tbb_found()) {
      assign("WINDOWS_PATH_SET", TRUE, envir = .GlobalEnv)
    } else {
      tbb_path <- file.path(
        get_tinystan_path(),
        "stan",
        "lib",
        "stan_math",
        "lib",
        "tbb"
      )
      Sys.setenv(PATH = paste(tbb_path, Sys.getenv("PATH"), sep = ";"))
      assign("WINDOWS_PATH_SET", tbb_found(), envir = .GlobalEnv)
    }
  }
}
